plugins/node/opentelemetry-instrumentation-pulsar-client/src/proxies/consumer.ts

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type * as Pulsar from 'pulsar-client';
+import type {ConsumerConfig} from 'pulsar-client';
+import * as api from '@opentelemetry/api';
+import {Attributes, Span, SpanStatusCode, Tracer} from '@opentelemetry/api';
+import {Instrumentation} from '../instrumentation';
+import {SemanticAttributes} from "@opentelemetry/semantic-conventions";
+import {PulsarInstrumentationConfig} from "../types";
+
+const spanNames = {
+  beforeConsume: 'beforeConsume',
+  afterConsume: 'afterConsume',
+};
+
+type ConsumerListener = (
+  message: Pulsar.Message,
+  consumer: Pulsar.Consumer
+) => void | Promise<void>;
+
+export class ConsumerProxy implements Pulsar.Consumer {
+  private readonly _tracer: Tracer;
+  private readonly _instrumentationConfig: PulsarInstrumentationConfig;
+  private readonly _moduleVersion: string | undefined;
+  private readonly config: Pulsar.ConsumerConfig;
+  private consumer: Pulsar.Consumer;
+
+  private _lastSpan: Span | undefined;
+  private _lastAttributes: Attributes | undefined;
+
+  constructor(
+    _tracer: Tracer,
+    _instrumentationConfig: PulsarInstrumentationConfig,
+    _moduleVersion: string | undefined,
+    config: Pulsar.ConsumerConfig,
+    consumer: Pulsar.Consumer
+  ) {
+    this._tracer = _tracer;
+    this._instrumentationConfig = _instrumentationConfig;
+    this._moduleVersion = _moduleVersion;
+    this.config = config;
+    this.consumer = consumer;
+  }
+
+  async receive(timeout?: number): Promise<Pulsar.Message> {
+    this.closePreviousSpan();
+    const message = await this.consumer.receive(timeout);
+
+    // Postpone the span ending for the next time the user calls receive
+    this._lastSpan = extractSpanFromMessage(
+      this._tracer,
+      this._instrumentationConfig,
+      this._moduleVersion,
+      this.config,
+      message
+    );
+    this._lastAttributes = getAttributesFromMessage(message);
+    return message;
+  }
+
+  private closePreviousSpan() {
+    // User is done with the last message and called receive again.
+    if (this._lastSpan) {
+      this._lastSpan.end();
+      if (this._instrumentationConfig.trackBeforeAndAfterConsume) {
+        trackBeforeAndAfterConsume(this._tracer, spanNames.afterConsume, this._lastAttributes!);
+      }
+    }
+  }
+
+  acknowledge(message: Pulsar.Message): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.acknowledge(message);
+  }
+
+  acknowledgeId(messageId: Pulsar.MessageId): Promise<null> {
+    this.closePreviousSpan();
```

Comment:
We have the id of the message being acknowledged, and we know the id of the message in `this._lastSpan` we will be closing in `closePreviousSpan`. I guess it can lead to using the wrong signal to end irrelevant spans, right? thus making the span duration not very useful or even misleading.